src/common/interop/KeyboardHook.cpp

```diff
@@ -0,0 +1,96 @@
+#include "pch.h"
+#include "KeyboardHook.h"
+#include <exception>
+#include <msclr\marshal.h>
+#include <msclr\marshal_cppstd.h>
+
+using namespace interop;
+using namespace System::Runtime::InteropServices;
+using namespace System;
+using namespace System::Diagnostics;
+
+KeyboardHook::KeyboardHook(
+    KeyboardEventCallback ^ keyboardEventCallback,
+    IsActiveCallback ^ isActiveCallback,
+    FilterKeyboardEvent ^ filterKeyboardEvent)
+{
+    kbEventDispatch = gcnew Thread(gcnew ThreadStart(this, &KeyboardHook::DispatchProc));
+    queue = gcnew Queue<KeyboardEvent ^>();
+    this->keyboardEventCallback = keyboardEventCallback;
+    this->isActiveCallback = isActiveCallback;
+    this->filterKeyboardEvent = filterKeyboardEvent;
+}
+
+KeyboardHook::~KeyboardHook()
+{
+    quit = true;
+    kbEventDispatch->Join();
+
+    // Unregister low level hook procedure
+    UnhookWindowsHookEx(hookHandle);
+}
+
+void KeyboardHook::DispatchProc()
+{
+    Monitor::Enter(queue);
+    quit = false;
+    while (!quit)
+    {
+        if (queue->Count == 0)
+        {
+            Monitor::Wait(queue);
+            continue;
+        }
+        auto nextEv = queue->Dequeue();
+
+        // Release lock while callback is being invoked
+        Monitor::Exit(queue);
+        
+        keyboardEventCallback->Invoke(nextEv);
+        
+        // Re-aquire lock
+        Monitor::Enter(queue);
+    }
+
+    Monitor::Exit(queue);
+}
+
+void KeyboardHook::Start()
+{
+    hookProc = gcnew HookProcDelegate(this, &KeyboardHook::HookProc);
+    Process ^ curProcess = Process::GetCurrentProcess();
+    ProcessModule ^ curModule = curProcess->MainModule;
+    // register low level hook procedure
+    hookHandle = SetWindowsHookEx(
+        WH_KEYBOARD_LL, 
+        (HOOKPROC)(void*)Marshal::GetFunctionPointerForDelegate(hookProc), 
+        GetModuleHandle(msclr::interop::marshal_as<std::wstring>(curModule->ModuleName).c_str()),
```

Comment:
Any reason why we cant pass in NULL like we do in existing code where we use SetWindowsHookEx?

Reply:
After looking at the documentation more closely, I think you are right, this should be set to null since the hook procedure is within the process and not in a different dll. https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexa